src/Microsoft.Diagnostics.Monitoring/ReversedServerConnectionsSource.cs

```diff
@@ -0,0 +1,194 @@
+
+        /// <summary>
+        /// Constructs a <see cref="ReversedServerConnectionsSource"/> that aggreates diagnostics connection
+        /// from a reversed diagnostics server at path specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        public ReversedServerConnectionsSource(string transportPath)
+        {
+            _server = new ReversedDiagnosticsServer(transportPath);
+
+            _acceptTask = AcceptAsync(_cancellation.Token);
+        }
+
+        public async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                _cancellation.Cancel();
+
+                _server.Dispose();
+
+                await _acceptTask;
+
+                _cancellation.Dispose();
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Gets the list of <see cref="ReversedDiagnosticsConnection"/> served from the reversed diagnostics server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>A list of active <see cref="IDiagnosticsConnection"/> instances.</returns>
+        public async Task<IEnumerable<IDiagnosticsConnection>> GetConnectionsAsync(CancellationToken token)
+        {
+            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _cancellation.Token);
+            CancellationToken linkedToken = linkedSource.Token;
+
+            // Prune connections that no longer have an active runtime instance before
+            // returning the list of connections.
+            await _connectionsSemaphore.WaitAsync(linkedToken);
+            try
+            {
+                // Create a task that checks each connection and removes it
+                // if it is no longer connected or is not responsive.
+                IList<Task> pruneTasks = _connections
+                    .Select(c => Task.Run(() => PruneConnectionAsync(c, linkedToken), linkedToken))
+                    .ToList();
+
+                await Task.WhenAll(pruneTasks);
+
+                return _connections
+                    .Select(c => new DiagnosticsConnection(c))
+                    .ToList()
+                    .AsReadOnly();
+            }
+            finally
+            {
+                _connectionsSemaphore.Release();
+            }
+        }
+
+        /// <summary>
+        /// Accepts connections from the reversed diagnostics server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        private async Task AcceptAsync(CancellationToken token)
```

Comment:
Minor nit: When I read `Accept*` I expect the method to return a connection.  You used `AcceptAsync` in other parts of the API in this manner, which maps closely to how it is used outside .NET.  Maybe this could be `Listen` instead?